src/Functions/UTCTimestampTransform.cpp

```diff
@@ -0,0 +1,169 @@
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 2nd argument type must be string.", name);
+            DataTypePtr date_time_type;
+            if (which_type_first.isDateTime())
+                date_time_type = std::make_shared<DataTypeDateTime>();
+            else
+            {
+                const DataTypeDateTime64 * date_time_64 = static_cast<const DataTypeDateTime64 *>(arguments[0].get());
+                date_time_type = std::make_shared<DataTypeDateTime64>(date_time_64->getScale());
+            }
+            return date_time_type;
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t) const override
+        {
+            if (arguments.size() != 2)
+                throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {}'s arguments number must be 2.", name);
+            ColumnWithTypeAndName arg1 = arguments[0];
+            ColumnWithTypeAndName arg2 = arguments[1];
+            const auto *  time_zone_col = checkAndGetColumn<ColumnString>(arg2.column.get());
+            if (!time_zone_col)
+            {
+                const auto * time_zone_const_col = checkAndGetColumn<ColumnConst>(arg2.column.get());
+                if (!time_zone_const_col)
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 2nd argument is not string type or const string type.", name);
+                time_zone_col = checkAndGetColumn<ColumnString>(time_zone_const_col->getDataColumn());
+                if (!time_zone_col)
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 2nd argument must be string type or const string type.", name);
+            }
+            auto column = result_type->createColumn();
+            if (WhichDataType(arg1.type).isDateTime())
+            {
+                const auto * date_time_col = checkAndGetColumn<ColumnDateTime>(arg1.column.get());
+                if (!date_time_col)
+                {
+                    const auto * date_time_const_col = checkAndGetColumn<ColumnConst>(arg1.column.get());
+                    if (!date_time_const_col)
+                        throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 1st argument is not datatime type or const type.", name);
+                    date_time_col = checkAndGetColumn<ColumnDateTime>(date_time_const_col->getDataColumn());
+                    if (!date_time_col)
+                        throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 1st argument must be datetime type", name);
+                }
+                for (size_t i = 0; i < date_time_col->size(); ++i)
+                {
+                    Field date_time_val;
+                    Field time_zone_val;
+                    date_time_col->get(i, date_time_val);
+                    time_zone_col->get(i, time_zone_val);
```

Comment:
Working with Fields is really slow and not optimal. Use concrete types and column methods `getDataAt`/`getElement`